### PR TITLE
add network policies

### PR DESCRIPTION
**Pull request by neo-liang-sap**
**What this PR does / why we need it**:
add network policies for following direction:
 - nginx-controller - from/to all

**Which issue(s) this PR fixes**:
Ref #565 (part of fixing)

**Special notes for your reviewer**:
cc @dansible 

**Release note**:

```noteworthy user
Every shoot cluster does now contain `NetworkPolicies` that allow the necessary traffic directions for the nginx-ingress-controller, metrics-server, and node-exporter.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Thanks @neo-liang-sap and @dansible! :) Have you tried this out already in a shoot?
Also, can you please enhance the release note a bit to explain more what exactly is added?

@mvladev can you please take a look here?

**Comment by neo-liang-sap** [comment]
Hi @rfranzke @mvladev 
I edited the release notes and i found some issue on my PR, will fix soon
Thanks.
-Neo

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    origin: gardener
+spec:
+  podSelector:
+    matchLabels:
+      app: nginx-ingress
+  policyTypes:
+  - Ingress
+  - Egress
+  egress:
+  - to:
```
Comment (mvladev) [inline charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml]: I would also add all `Pods` to all namepaces to the egress as well.

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    - ipBlock:
+        cidr: 0.0.0.0/0
+  ingress:
```
Comment (mvladev) [inline charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml:20]: Allow ingress from all Pods and from all namespaces. For example if Pod-A a tries to access a service via its LoadBalancer IP, if there is no policy allowing that Pod-A is now allowed at the ingress, it won't work.

File: charts/shoot-core/components/charts/network-policies/templates/metrics-server-from-vpn.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  podSelector:
+    matchLabels:
+      k8s-app: metrics-server
+  policyTypes:
+  - Ingress
+  ingress:
+  - from:
+    - podSelector:
+        matchLabels:
+          app: vpn-shoot
```
Comment (mvladev) [inline charts/shoot-core/components/charts/network-policies/templates/metrics-server-from-vpn.yaml]: There is a policy which already exists - https://github.com/gardener/gardener/blob/4dab9070a5ece73b2a6ad1a585a036c4a410108b/charts/shoot-core/components/charts/network-policies/templates/allow-from-seed.yaml#L1-L24

Label applicable pods with `networking.gardener.cloud/from-seed: allowed` instead

File: charts/shoot-core/components/charts/network-policies/templates/node-exporter-to-api-server.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  podSelector:
+    matchLabels:
+      component: node-exporter
+  policyTypes:
+  - Egress
+  egress:
+  - to:
+    - podSelector:
+        matchLabels:
+          role: apiserver
```
Comment (mvladev) [inline charts/shoot-core/components/charts/network-policies/templates/node-exporter-to-api-server.yaml]: Node exporter talks via the API server on it's public IP.  Use https://github.com/gardener/gardener/blob/4dab9070a5ece73b2a6ad1a585a036c4a410108b/charts/shoot-core/components/charts/network-policies/templates/allow-to-public-networks.yaml#L1-L21 instead and label this pod with `networking.gardener.cloud/to-public-networks: allowed`

**Review by mvladev — CHANGES_REQUESTED** [review]

**Comment by neo-liang-sap** [comment]
Hi @rfranzke , i tested in my local shoot and corrected some typo - sorry for this and thanks for the catch!
Hi @mvladev , i added `- namespaceSelector: {}` to select all pods in all namespaces, do correct me if i was wrong...Thanks!
```
  - to:
    - ipBlock:
        cidr: 0.0.0.0/0
    - namespaceSelector: {}
```

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml (hunk 1 above)
Comment (mvladev) [reply charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml]: Actually 

```yaml
ingress: 
- {}
```
 should be sufficient. https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-allow-all-ingress-traffic

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml (hunk 2 above)
Comment (mvladev) [reply charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml:20]:
```yaml
egress:
- {}
```
should be sufficient https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-deny-all-ingress-and-all-egress-traffic

**Comment by neo-liang-sap** [comment]
Hi @rfranzke , i've tested in my local shoot, sorry for my typo and thanks for catching!
Hi @mvladev , thanks for the comments, i've modified according to your suggest and do let me know anything wrong, thanks!

-Neo

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  annotations:
+    gardener.cloud/description: |
+      Allows Egress and Ingress from the nginx-controller to all traffice
```
Comment (danielfoehrKn) [inline charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml]:
```suggestion
      Allows all Egress and Ingress for the nginx-controller
```

Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml]: thanks @danielfoehrKn ! i've corrected as you commented.
-Neo

**Comment by ialidzhikov** [comment]
> Add Network Policy for dashboard to API server

I think there is no dashboard related change in this PR, right? Can you update the PR description?

**Comment by neo-liang-sap** [comment]
Hi @ialidzhikov , i've changed description according to your comments, thanks!

File: charts/shoot-core/components/charts/network-policies/templates/node-exporter-from-vpn.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+kind: NetworkPolicy
```
Comment (mvladev) [inline charts/shoot-core/components/charts/network-policies/templates/node-exporter-from-vpn.yaml]: I just had a view and there is already another policy https://github.com/mvladev/gardener/blob/4dab9070a5ece73b2a6ad1a585a036c4a410108b/charts/shoot-core/components/charts/network-policies/templates/allow-from-seed.yaml#L1-L24

which allows for Pods to be labeled with `networking.gardener.cloud/from-seed: allowed`.
So this NetworkPolicy would not be needed and you can simply label the node-exporter with `networking.gardener.cloud/from-seed: allowed`

Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/network-policies/templates/node-exporter-from-vpn.yaml]: thanks! i've modified per comments

File: charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml [hunk 7]
```diff
@@ -29,6 +29,8 @@ metadata:
   name: node-exporter
   namespace: kube-system
   labels:
+    networking.gardener.cloud/to-public-networks: allowed
```
Comment (mvladev) [inline charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: Node exporter doesn't need to talk to any component? I don't think this is needed

File: charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml [hunk 8]
```diff
@@ -5,6 +5,7 @@ metadata:
   name: metrics-server
   namespace: kube-system
   labels:
+    networking.gardener.cloud/from-seed: allowed
```
Comment (mvladev) [inline charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml]: Metric server would also need DNS access as well.
Also access to the Nodes CIDR

File: charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml (hunk 7 above)
Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: this line was added per your previous comments https://github.com/gardener/gardener/pull/2044#discussion_r392054792
Thanks!

File: charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml (hunk 8 above)
Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml]: thanks, i've updated the file

File: charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml (hunk 7 above)
Comment (mvladev) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: I was mistaken I meant `metric server`... Sorry for the confusion.

Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: Hi @mvladev , currently in `metrics-server\deployment.yaml` there are two network policies
```
    networking.gardener.cloud/from-seed: allowed
    networking.gardener.cloud/to-dns: allowed
```
Sorry i'm confused do i still need to modify this file?
Thanks.

Comment (mvladev) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: ` networking.gardener.cloud/to-dns: allowed` is not needed.

To summarize: 

1. node-exporter needs to receive traffic only from the VPN
1. metric server needs to receive traffic only from the VPN
1. metric server needs to send traffic to the DNS and the API server.

Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: thanks - i've removed this line

**Comment by neo-liang-sap** [comment]
Hello,
I just commit one new [network policy](https://github.com/gardener/gardener/pull/2044/commits/ce6171f3b96b3e2ae673bcc5b73eefab10da829e) for calico to api server in TCP 443 port, for more detailed info please refer to discussion on https://github.com/gardener/gardener-extension-networking-calico/pull/18#issuecomment-600591743 , to fix https://github.com/gardener/gardener-extension-networking-calico/issues/16
Then i will label calico related components with this network policy
Thanks.
-Neo

File: charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  annotations:
+    gardener.cloud/description: |
+      Allows calico related traffic to api server in TCP Port 443
+  name: gardener.cloud--allow-calico-to-apiserver
```
Comment (zanetworker) [inline charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml]:
```suggestion
  name: gardener.cloud--allow-to-apiserver
```

File: charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  namespace: kube-system
+  labels:
+    origin: gardener
+spec:
+  podSelector:
+    matchLabels:
+      networking.gardener.cloud/calico-to-apiserver: allowed
```
Comment (zanetworker) [inline charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml]:
```suggestion
      networking.gardener.cloud/to-apiserver: allowed
```

**Comment by neo-liang-sap** [comment]
thanks @zanetworker , i've fixed the np name

File: charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+kind: NetworkPolicy
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml]: Can you rename the file to `allow-to-api-server.yaml`?

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
+kind: NetworkPolicy
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml:1]: I don't see this network policy is used by the nginx-ingress-controller yet?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml (hunk 11 above)
Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/network-policies/templates/allow-calico-to-api-server.yaml]: thanks, i've renamed the file

File: charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml (hunk 12 above)
Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml:1]: i use following pod selector in this network policy file 
```
spec:
  podSelector:
    matchLabels:
      app: nginx-ingress
```

Comment (rfranzke) [reply charts/shoot-core/components/charts/network-policies/templates/allow-from-to-nginx.yaml:1]: I see, thanks 👍

**Review by rfranzke — COMMENTED** [review]
/lgtm, @mvladev @zanetworker can you have another check as well, please?

**Review by zanetworker — DISMISSED** [review]
/lgtm

**Review by mvladev — CHANGES_REQUESTED** [review]
Hold on they are not correct - metrics server still need to talk to the API server and also to the nodes (which needs DNS)

**Comment by zanetworker** [comment]
@mvladev isn't the metrics server directly talking to the kubelet, it just needs to be reachable from the API server?

**Comment by mvladev** [comment]
https://github.com/kubernetes-sigs/metrics-server/blob/65b8126df5b4037162e3c696c1216ffea67aaff4/pkg/api/node.go#L74-L110

It needs to get information about Nodes and Pods by calling the apiserver

**Comment by neo-liang-sap** [comment]
Hi @mvladev , @zanetworker ,
i've added two np to metrics server so totally there are 3 nps for metrics server, thanks
```
    networking.gardener.cloud/from-seed: allowed
    networking.gardener.cloud/to-apiserver: allowed
    networking.gardener.cloud/to-dns: allowed
```

**Comment by mvladev** [comment]
metrics-server talks to the kubelet on port `10250` and the current policy would block this. See: 
https://github.com/kubernetes-sigs/metrics-server/blob/65b8126df5b4037162e3c696c1216ffea67aaff4/cmd/metrics-server/app/options/options.go#L93

**Comment by neo-liang-sap** [comment]
@mvladev should i add one new networkpolicy like `networking.gardener.cloud/to-kubelet` for port `10250`?

**Comment by mvladev** [comment]
> @mvladev should i add one new networkpolicy like networking.gardener.cloud/to-kubelet for port 10250?

Yes, it makes sense.

**Comment by zanetworker** [comment]
Also just a note this configuration is tied to a flag and not a fixed value, which means, it needs to be changed if the flag changes `kubelet-port`.

**Comment by mvladev** [comment]
> Also just a note this configuration is tied to a flag and not a fixed value, which means, it needs to be changed if the flag changes kubelet-port.

It doesn't matter for our use - case. It will change only if `kubelet` changes the default port (it won't). Also we have full control over this.

**Comment by neo-liang-sap** [comment]
how can i get this `kubelet-port` value like an variable? @zanetworker @mvladev

**Comment by zanetworker** [comment]
@neo-liang-sap for now just keep it as is I would say.

**Comment by neo-liang-sap** [comment]
i added an np for kubelet and label metrics server with it

File: charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml [hunk 13]
```diff
@@ -5,6 +5,10 @@ metadata:
...
+    networking.gardener.cloud/to-apiserver: allowed
+    networking.gardener.cloud/to-dns: allowed
+    networking.gardener.cloud/to-kubelet: allowed
```
Comment (mvladev) [inline charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml]: All those labels must be add to the Pod template several lines down. This applies for all resources. Adding them to the Deployment label would have no effect

Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml]: done, thanks

Comment (mvladev) [reply charts/shoot-core/components/charts/metrics-server/templates/deployment.yaml]: They are not needed in `Deployment.metadata.labels` they must only be in `Deployment.spec.template.metadata.labels`. Remove them here.

File: charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml [hunk 14]
```diff
@@ -29,6 +29,8 @@ metadata:
...
+    networking.gardener.cloud/from-seed: allowed
```
Comment (mvladev) [inline charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: They are not needed in `Demonset.metadata.labels` they must only be in `Demonset.spec.template.metadata.labels`. Remove them here.

**Review by mvladev — CHANGES_REQUESTED** [review]

File: charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml (hunk 14 above)
Comment (neo-liang-sap) [reply charts/shoot-core/components/charts/monitoring/charts/node-exporter/templates/node-exporter.yaml]: done with both changes, thanks

File: charts/shoot-core/components/charts/network-policies/templates/allow-to-api-server.yaml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
```
Comment (zanetworker) [inline charts/shoot-core/components/charts/network-policies/templates/allow-to-api-server.yaml]: missing  kind: NetworkPolicy

**Review by mvladev — DISMISSED** [review]

**Review by rfranzke — COMMENTED** [review]
/lgtm, played with it on GCP and it seems all affected components (nginx-ingress, metrics-server, node-exporter) still work as expected.

@neo-liang-sap can you please squash the commits into a single one so that we can get this PR merged?

**Comment by neo-liang-sap** [comment]
@rfranzke , thanks! i've squash into one commit.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
Thank you @neo-liang-sap for your contribution and all your efforts! :)

**Comment by neo-liang-sap** [comment]
welcome @rfranzke !

**Pull request merged**